pkg/sethandler/controller.go

```diff
@@ -113,35 +122,55 @@ func (setHandler *SetHandler) PodChanged(oldPod, newPod v1.Pod) {
 	if !shouldPodBeHandled(newPod) {
 		return
 	}
-	setHandler.adjustContainerSets(newPod)
+	containersToBeSet := map[string]int{}
+	if newPod.ObjectMeta.Annotations[setterAnnotationKey] != "" {
+		containersToBeSet = determineContainersToBeSet(oldPod, newPod)
+	} else {
+		for _, containerStatus := range newPod.Status.ContainerStatuses {
+			if containerStatus.ContainerID == "" {
+				return
+			}
+			containersToBeSet[containerStatus.Name] = 0
+		}
+	}
+	if len(containersToBeSet) > 0 {
+		setHandler.adjustContainerSets(newPod, containersToBeSet)
+	}
 }
 
 func shouldPodBeHandled(pod v1.Pod) bool {
 	setterNodeName := os.Getenv("NODE_NAME")
 	podNodeName := pod.Spec.NodeName
-
 	//Pod is not yet scheduled, or it wasn't scheduled to the Node of this specific CPUSetter instance
 	if setterNodeName == "" || podNodeName == "" || setterNodeName != podNodeName {
 		return false
 	}
-	//If the Pod is not running, its containers haven't been created yet - no cpuset cgroup is present to be adjusted by the CPUSetter
-	if pod.ObjectMeta.Annotations[setterAnnotationKey] != "" {
-		return false
-	}
-	if len(pod.Status.ContainerStatuses) == 0 {
-		return false
-	}
-	for _, containerStatus := range pod.Status.ContainerStatuses {
-		if containerStatus.ContainerID == "" {
-			return false
+	return true
+}
+
+func determineContainersToBeSet(oldPod, newPod v1.Pod) map[string]int {
+	workingContainers := map[string]int{}
+	found := false
+	for _, newContainerStatus := range newPod.Status.ContainerStatuses {
+		found = false
+		for _, oldContainerStatus := range oldPod.Status.ContainerStatuses {
+			if oldContainerStatus.ContainerID == newContainerStatus.ContainerID {
+				found = true
+			}
+		}
+		if !found {
+			workingContainers[newContainerStatus.Name] = 0
```

Comment:
are you sure Names are distinct? shouldn't we use ID instead?

Reply:
"name string | This must be a DNS_LABEL. Each container in a pod must have a unique name. Cannot be updated."
ok I guess we can be sure they are unique, so feel free to ignore